ome/datasource_application_certificate.go

```diff
@@ -0,0 +1,153 @@
+/*
```

Comment:
```
+++ README.md
@@ 1 +2 @@
-Unless required by applicable law or agreed to in writing, software
+during Terraform provider development, ensure that all code adheres to the Hasicorp Terraform guidelines for readability and maintainability.

+++ file.go
@@ 1,3 +1,5 @@
+package ome
+
+import (
	"context"
	"terraform-provider-ome/models"

	"github.com/hashicorp/terraform-plugin-framework/datasource"
	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
)

var (
	_ datasource.DataSource              = &appCertDatasource{}
	_ datasource.DataSourceWithConfigure = &appCertDatasource{}
)

// NewAppCertDataSource is new datasource for application certificate
func NewAppCertDataSource() datasource.DataSource {
	return &appCertDatasource{}
}

+++ appCertDatasource.go
@@ -10 +10 @@
+type appCertDatasource struct {
+	p *omeProvider
+}

+++ appCertDatasource.go
@@ -24 +24 @@
+// Schema implements datasource.DataSource
+func (g *appCertDatasource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "Data source to read application certificate data from OpenManage Enterprise.",
+		Attributes: map[string]schema.Attribute{
+			// Use type constraints for attribute values
+			"id": schema.StringAttribute{
+				MarkdownDescription: "ID for application certificate data source.",
+				Description:         "ID for application certificate data source.",
+				Computed:            true,
+				Optional:            true,
+			},
+			"issued_to": schema.SingleNestedAttribute{
+				MarkdownDescription: "List of the device id(s) associated with any of the groups.",
+				Description:         "List of the device id(s) associated with any of the groups.",
+				Attributes:          g.infoSchema(),
+				Computed:            true,
+			},
+			"issued_by": schema.SingleNestedAttribute{
+				MarkdownDescription: "List of the device servicetags associated with any of the groups.",
+				Description:         "List of the device servicetags associated with any of the groups.",
+				Attributes:          g.infoSchema(),
+				Computed:            true,
+			},
+			// Use type constraints for attribute values
+			"valid_to": schema.StringAttribute{
+				MarkdownDescription: "List of the device group names.",
+				Description:         "List of the device group names.",
+				Computed:            true,
+			},
+			"valid_from": schema.StringAttribute{
+				MarkdownDescription: "Map of the groups fetched keyed by its name.",
+				Description:         "Map of the groups fetched keyed by its name.",
+				Computed:            true,
+			},
+		},
+	}
+}

+++ appCertDatasource.go
@@ -82,7 +82,8 @@
+	"email": schema.StringAttribute{
+		MarkdownDescription: "Email address.",
+		Description:         "Email address.",
+		Computed:            true,
+	},
```

Report:
- The `README.md` file should include a section on the guidelines for Terraform provider development to ensure that all code adheres to the Hasicorp Terraform guidelines for readability and maintainability.
- In the `appCertDatasource` struct, the `p` field should be initialized with a new `omeProvider` instance when the `appCertDatasource` is created.
- The `Schema` function in the `appCertDatasource` should use type constraints for attribute values, such as `schema.StringAttribute` and `schema.SingleNestedAttribute`, to ensure data type consistency.
- The `infoSchema` function should be refactored to use type constraints for attribute values, such as `schema.StringAttribute` and `schema.SingleNestedAttribute`, to ensure data type consistency.